File: Utilities/Matlab/scripts/ns2d.m
% Roger Wang
% 6-28-11
% ns2d.m
% Process the output of the ns2d verification cases

close all
clear all

infile{1,1} = '../../Verification/NS_Analytical_Solution/ns2d_8_devc.csv';
infile{2,1} = '../../Verification/NS_Analytical_Solution/ns2d_16_devc.csv';
infile{3,1} = '../../Verification/NS_Analytical_Solution/ns2d_32_devc.csv';
infile{4,1} = '../../Verification/NS_Analytical_Solution/ns2d_64_devc.csv';
infile{1,2} = '../../Verification/NS_Analytical_Solution/ns2d_8_nupt1_devc.csv';
infile{2,2} = '../../Verification/NS_Analytical_Solution/ns2d_16_nupt1_devc.csv';
infile{3,2} = '../../Verification/NS_Analytical_Solution/ns2d_32_nupt1_devc.csv';
infile{4,2} = '../../Verification/NS_Analytical_Solution/ns2d_64_nupt1_devc.csv';

outfile{1,1} = '../../Verification/NS_Analytical_Solution/ns2d_8_exact.csv';
outfile{2,1} = '../../Verification/NS_Analytical_Solution/ns2d_16_exact.csv';
outfile{3,1} = '../../Verification/NS_Analytical_Solution/ns2d_32_exact.csv';
outfile{4,1} = '../../Verification/NS_Analytical_Solution/ns2d_64_exact.csv';
outfile{1,2} = '../../Verification/NS_Analytical_Solution/ns2d_8_nupt1_exact.csv';
outfile{2,2} = '../../Verification/NS_Analytical_Solution/ns2d_16_nupt1_exact.csv';
outfile{3,2} = '../../Verification/NS_Analytical_Solution/ns2d_32_nupt1_exact.csv';
outfile{4,2} = '../../Verification/NS_Analytical_Solution/ns2d_64_nupt1_exact.csv';

nu(1) = 0.0;
nu(2) = 0.1;

pi = 4.0*atan(1.0);
x = pi;
y = pi;

dx(1) = 2.0*pi/8.0;
dx(2) = 2.0*pi/16.0;
dx(3) = 2.0*pi/32.0;
dx(4) = 2.0*pi/64.0;

for  k=1:2
    for j=1:4
        M_11 = fopen(outfile{j,k}, 'wt');
        fprintf(M_11, '%s\n', 'Time, u-vel');
        if ~exist(infile{j,k})
            display(['Error: File ',infile{j,k},' does not exist. Skipping case.'])
            return
        end
        M_10 = csvread(infile{j,k}, 2, 0);
        rms(j,k) = 0.0;
        cnt(j,k) = 0.0;
        for i = 1:length(M_10(:,1))
            t(i) = M_10(i,1);
            u(i,j,k)= M_10(i,2);
            u_exact = 1.0 - 2.0*cos(x-t(i))*sin(y-0.5*dx(j)-t(i))*exp(-2.0*nu(k)*t(i));
            fprintf(M_11,'%8.3f, %9.5f\n', t(i), u_exact);
            rms(j,k) = rms(j,k) + (u(i,j,k)-u_exact).^2;
            cnt(j,k) = cnt(j,k) + 1.0;
        end
        fclose(M_11);
        rms(j,k) = sqrt(rms(j,k)/cnt(j,k));
    end
end

       
%Write the error files
dir = '../../Verification/NS_Analytical_Solution/';

filename11 = [dir,'ns2d_error.csv'];
fid11 = fopen(filename11,'wt');
filename12 = [dir,'ns2d_nupt1_error.csv'];
fid12 = fopen(filename12,'wt');

fprintf(fid11,'%s\n', 'dx,dx^2,rms error');
fprintf(fid12,'%s\n', 'dx,dx^2,rms error');
   
for j = 1:4
    fprintf(fid11,'%8.3f, %8.3f, %8.4f\n', dx(j),dx(j)^2, rms(j,1));
    fprintf(fid12,'%8.3f, %8.3f, %8.4f\n', dx(j),dx(j)^2, rms(j,2));
end
fclose(fid11);
fclose(fid12);

File: Utilities/Python/scripts/ns2d.py
# Roger Wang
# 6-28-11
# ns2d.py
# Process the output of the ns2d verification cases
#
# Converted by Floyd
# 10/17/2025

import numpy as np
import pandas as pd
import os
import math


datadir = '../../Verification/NS_Analytical_Solution/'
infile = [['ns2d_8_devc.csv','ns2d_8_nupt1_devc.csv'],
   ['ns2d_16_devc.csv','ns2d_16_nupt1_devc.csv'],
   ['ns2d_32_devc.csv','ns2d_32_nupt1_devc.csv'],
   ['ns2d_64_devc.csv','ns2d_64_nupt1_devc.csv']]

outfile = [['ns2d_8_exact.csv','ns2d_8_nupt1_exact.csv'],
   ['ns2d_16_exact.csv','ns2d_16_nupt1_exact.csv'],
   ['ns2d_32_exact.csv','ns2d_32_nupt1_exact.csv'],
   ['ns2d_64_exact.csv','ns2d_64_nupt1_exact.csv']]

skip_case = False

for i in range(len(infile)):
   for j in range(len(infile[i])):
      name = datadir+infile[i][j]
      if not os.path.exists(name):
         skip_case = True
         print('Error: File ', infile[i][j], ' does not exist. Skipping case.')

if skip_case: quit()

nu = [0.0, 0.1]
pi = math.pi
x = pi
y = pi
dx = [2.0 * pi / 8.0, 2.0 * pi / 16.0, 2.0 * pi / 32.0, 2.0 * pi / 64.0]

rms = np.zeros((4, 2))
cnt = np.zeros((4, 2))

# k iterates over viscosity (0: nu=0.0, 1: nu=0.1)
for k in range(2):
   # j iterates over resolution (0: 8 cells, 3: 64 cells)
   for j in range(4):
      M_10 = pd.read_csv(datadir+infile[j][k], skiprows=2, header=None).values

      with open(datadir+outfile[j][k], 'w') as M_11:
         M_11.write("Time, u-vel\n")

         squared_error_sum = 0.0
         num_points = 0.0

         for i in range(len(M_10)):
            t = M_10[i, 0]
            u_fds = M_10[i, 1]

            u_exact = 1.0 - 2.0 * math.cos(x - t) * math.sin(y - 0.5 * dx[j] - t) * math.exp(-2.0 * nu[k] * t)

            M_11.write(f"{t:8.3f}, {u_exact:9.5f}\n")

            squared_error_sum += (u_fds - u_exact)**2

            num_points += 1.0

         if num_points > 0:
            rms[j, k] = math.sqrt(squared_error_sum / num_points)
         else:
            rms[j, k] = 0.0

# Case 1: nu=0.0 (k=0)
with open(datadir+'ns2d_error.csv', 'w') as fid11:
   fid11.write('dx,dx^2,rms error\n')
   for j in range(4):
      fid11.write(f"{dx[j]:8.3f}, {dx[j]**2:8.3f}, {rms[j, 0]:8.4f}\n")

# Case 2: nu=0.1 (k=1)
with open(datadir+'ns2d_nupt1_error.csv', 'w') as fid12:
   fid12.write('dx,dx^2,rms error\n')
   for j in range(4):
      fid12.write(f"{dx[j]:8.3f}, {dx[j]**2:8.3f}, {rms[j, 1]:8.4f}\n")

